### PR TITLE
[feature](nereids)support distinct count

### PR DESCRIPTION
**Pull request by yinzhijian**
# Proposed changes

Issue Number: close #xxx

## Problem summary
support distinct count with group by clause.
for example:
SELECT count(distinct c_custkey + 1) FROM customer group by c_nation;

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Count.java [hunk 1]
```diff
@@ -31,21 +31,28 @@
 public class Count extends AggregateFunction {
 
     private final boolean isStar;
+    private final boolean isDistinct;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Count.java:34]: move it to super class, and add a two parameters constructor. the original constructor set it to false by default. In super class, method isDistinct return this value directly

File: regression-test/data/nereids_syntax_p0/function.out [hunk 2]
```diff
@@ -11,6 +11,11 @@
 -- !count --
 3	3
 
+-- !distinct_count_with_group_by --
+1	1
+1	1
+1	1
```
Comment (morrySnow) [inline regression-test/data/nereids_syntax_p0/function.out:17]: we need to load more data to table to verify it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java [hunk 3]
```diff
@@ -54,86 +56,192 @@
...
-            return new LogicalAggregate<>(
-                    globalGroupByExprs,
-                    globalOutputExprs,
-                    true,
-                    aggregate.isNormalized(),
-                    AggPhase.GLOBAL,
-                    localAggregate
-            );
+    @Override
+    public Rule build() {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java:236]: move build as the first function of this class for easy reading

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java [hunk 4]
```diff
@@ -54,86 +56,192 @@
...
-            for (Expression originGroupByExpr : originGroupByExprs) {
-                if (inputSubstitutionMap.containsKey(originGroupByExpr)) {
+    private Map<Expression, Expression> inputSubstitutionMap = Maps.newHashMap();
+    private List<NamedExpression> distinctAggFunctionParams = new ArrayList<>();
+    private List<AggregateFunction> distinctAggFunctions = new ArrayList<>();
+    private List<NamedExpression> distinctOriginOutputExprs = new ArrayList<>();
+
+    // only support distinct function with group by
+    // TODO: support distinct function without group by. (add second global phase)
+    private LogicalAggregate secondDisassemble(LogicalAggregate<GroupPlan> aggregate) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java:66]: please add some UT for this class

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java [hunk 5]
```diff
@@ -54,86 +56,192 @@
...
+                    originOutputExpr.collect(AggregateFunction.class::isInstance);
+            for (AggregateFunction aggregateFunction : aggregateFunctions) {
+                if (aggregateFunction.isDistinct()) {
+                    distinctAggFunctions.add(aggregateFunction);
+                    distinctOriginOutputExprs.add(originOutputExpr);
                 }
             }
+        }
+        if (!distinctAggFunctions.isEmpty()) {
+            for (Expression expr : distinctAggFunctions.get(0).children()) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java:226]: why only get(0)? pls add some comments

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java [hunk 6]
```diff
@@ -54,86 +56,192 @@
...
+                distinctAggFunctionParams.add((NamedExpression) expr);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java:227]: why could assume expr is `NamedExpression`? how about `count(distinct a + 1)`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java [hunk 7]
```diff
@@ -54,86 +56,192 @@
...
+        // global agg: select a from t1 group by b, a;
+        // second local agg: select count(distinct a) from t1 group by b;
+        // In order to get the second local agg from global agg:
+        // 1. the distinct expression needs to be removed from the output and the group by of global agg
+        // 2. add distinct agg function back to output
+        List<NamedExpression> originFirstGlobalOutputExprs = aggregate.getOutputExpressions();
+        List<Expression> originFirstGlobalGroupByExprs = aggregate.getGroupByExpressions();
+        List<NamedExpression> secondLocalOutputExprs = new ArrayList<>(aggregate.getOutputExpressions());
+        // add origin distinct function back
+        secondLocalOutputExprs.addAll(distinctOriginOutputExprs);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java:77]: secondLocalOutputExprs generated from aggregate output expressions, i think it already has origin distinct function?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java [hunk 8]
```diff
@@ -54,86 +56,192 @@
...
+    // only support distinct function with group by
+    // TODO: support distinct function without group by. (add second global phase)
+    private LogicalAggregate secondDisassemble(LogicalAggregate<GroupPlan> aggregate) {
+        // origin sql: select count(distinct a) from t1 group by b;
+        // global agg: select a from t1 group by b, a;
+        // second local agg: select count(distinct a) from t1 group by b;
+        // In order to get the second local agg from global agg:
+        // 1. the distinct expression needs to be removed from the output and the group by of global agg
+        // 2. add distinct agg function back to output
+        List<NamedExpression> originFirstGlobalOutputExprs = aggregate.getOutputExpressions();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java:73]: we'd better use phase name same as AggPhase i.e.
- LOCAL
- GLOBAL
- DISTINCT_LOCAL
- DISTINCT_GLOBAL

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java [hunk 9]
```diff
@@ -54,86 +56,192 @@
  */
 public class AggregateDisassemble extends OneRewriteRuleFactory {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java:57]: pls update this class's java doc comments, add distinct situation

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 10]
```diff
@@ -230,6 +246,13 @@ public PlanFragment visitPhysicalAggregate(PhysicalAggregate<Plan> aggregate, Pl
             case GLOBAL:
                 inputPlanFragment.updateDataPartition(mergePartition);
                 return inputPlanFragment;
+            case DISTINCT_LOCAL:
+                AggregationNode globalAggNode = (AggregationNode) aggregationNode.getChild(0);
+                globalAggNode.unsetNeedsFinalize();
+                globalAggNode.setIntermediateTuple();
+                inputPlanFragment.updateDataPartition(mergePartition);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:253]: when we do distinct agg, the partition exprs is different from group by exprs, where to process this?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java [hunk 11]
```diff
@@ -54,86 +56,192 @@
...
+        List<NamedExpression> secondLocalOutputExprs = new ArrayList<>(aggregate.getOutputExpressions());
+        // add origin distinct function back
+        secondLocalOutputExprs.addAll(distinctOriginOutputExprs);
+        List<Expression> secondLocalGroupByExprs = new ArrayList<>(aggregate.getGroupByExpressions());
+
+        List<Expression> distinctExprs = new ArrayList<>();
+        // remove distinct param exprs from secondLocalGroupByExprs and secondLocalOutputExprs
+        for (NamedExpression expression : distinctAggFunctionParams) {
+            secondLocalGroupByExprs.remove(inputSubstitutionMap.get(expression));
+            secondLocalOutputExprs.remove(inputSubstitutionMap.get(expression));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AggregateDisassemble.java:84]: secondLocalOutputExprs has NamedExpression, such as Alias and SlotReference, but i don't think key of inputSubstitutionMap has the same Alias

File: regression-test/data/nereids_syntax_p0/function.out (hunk 2 above)
Comment (yinzhijian) [reply regression-test/data/nereids_syntax_p0/function.out]: Data cannot be easily added to the existing test set...

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java (hunk 10 above)
Comment (yinzhijian) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:225]: partition exprs same as DISTINCT_LOCAL group by exprs

**Review by morrySnow — APPROVED** [review]

**Pull request merged**